### PR TITLE
Add preprocessed output support

### PR DESCRIPTION
**Pull request by balazs-toth-ing**
This change extend the plugin functionality with optionally save the preprocessed output of the generated diagrams. The preprocessed diagram sources are useful for example to use on Confluence where some advanced feature of PlantUML is disabled in the render plugin and SVGs not allowed regards to security concerns (SVG scripting).

### CODE REVIEW

**Comment by davidmoten** [comment]
Thanks, I'll check it out

**Comment by balazs-toth-ing** [comment]
Please let me know if you have any questions or suggestions. Can you please estimate when you will take a look?

**Comment by davidmoten** [comment]
Thanks for the reminder, I'll have a look in the next couple of days.

File: plantuml-maven-plugin-test/pom.xml [hunk 1]
```diff
@@ -49,6 +49,7 @@
                                 <configFile>src/main/plantuml2/global.config</configFile>
                             </configFiles>
                             <metadata>true</metadata>
+							<storePreproc>true</storePreproc>
```
Comment (davidmoten) [inline plantuml-maven-plugin-test/pom.xml]: spaces not tabs please

File: README.md [hunk 2]
```diff
@@ -85,6 +85,7 @@ Defaults for the `<configuration>` element are:
 * **configs**: none
 * **formats**: `png`
 * **metadata**: `true`
+* **storePreproc**: `false`
```
Comment (davidmoten) [inline README.md]: I prefer `writePreproc`, thanks

File: plantuml-maven-plugin/src/main/java/com/github/davidmoten/plantuml/plugins/GenerateMojo.java [hunk 3]
```diff
@@ -52,6 +55,9 @@ public final class GenerateMojo extends AbstractMojo {
     @Parameter(name = "metadata", defaultValue = "true")
     private boolean metadata;
 
+    @Parameter(name = "storePreproc", defaultValue = "false")
```
Comment (davidmoten) [inline plantuml-maven-plugin/src/main/java/com/github/davidmoten/plantuml/plugins/GenerateMojo.java]: let's use `writePreproc` throughout

File: plantuml-maven-plugin/src/main/java/com/github/davidmoten/plantuml/plugins/GenerateMojo.java [hunk 4]
```diff
@@ -167,4 +176,22 @@ private static String commaSeparate(List<String> list) {
         return list.stream().collect(Collectors.joining(","));
     }
 
+	private void extractPreproc(SourceFileReaderExt sourceFileReader) throws IOException {
+		for (BlockUml blockUml : sourceFileReader.getBlocks()) {
+			final SuggestedFile suggested = sourceFileReader.extractSuggestedFile(blockUml)
+					.withPreprocFormat();
+			final SFile file = suggested.getFile(0);
+			getLog().info("Export preprocessing source to " + file.getPrintablePath());
```
Comment (davidmoten) [inline plantuml-maven-plugin/src/main/java/com/github/davidmoten/plantuml/plugins/GenerateMojo.java]: `Export` -> `exporting` please (note lower case for logging consistency for this project)

File: plantuml-maven-plugin/src/main/java/com/github/davidmoten/plantuml/plugins/SourceFileReaderExt.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import net.sourceforge.plantuml.preproc.Defines;
+
+public class SourceFileReaderExt extends SourceFileReader {
+
+	public SourceFileReaderExt(Defines defines, File file, File outputDirectory, List<String> config, String charset,
+			FileFormatOption fileFormatOption) throws IOException {
+		super(defines, file, outputDirectory, config, charset, fileFormatOption);
+	}
+
+	public SuggestedFile extractSuggestedFile(BlockUml blockUml) throws FileNotFoundException {
```
Comment (davidmoten) [inline plantuml-maven-plugin/src/main/java/com/github/davidmoten/plantuml/plugins/SourceFileReaderExt.java]: pointless use of inheritance here, please get rid of this class, just call `sourceFileReader.getSuggestedFile(blockUml)` from the method `extractPreproc`.

**Review by davidmoten — CHANGES_REQUESTED** [review]

File: plantuml-maven-plugin/src/main/java/com/github/davidmoten/plantuml/plugins/SourceFileReaderExt.java (hunk 5 above)
Comment (davidmoten) [reply plantuml-maven-plugin/src/main/java/com/github/davidmoten/plantuml/plugins/SourceFileReaderExt.java]: Ah I see that `getSuggestedFile` is a protected method so you had to use inheritance, no worries, I'll mark this as resolved.

File: plantuml-maven-plugin-test/pom.xml (hunk 1 above)
Comment (balazs-toth-ing) [reply plantuml-maven-plugin-test/pom.xml]: fixed (everywhere I touched the code, including the new java method in the mojo)

File: plantuml-maven-plugin/src/main/java/com/github/davidmoten/plantuml/plugins/SourceFileReaderExt.java (hunk 5 above)
Comment (balazs-toth-ing) [reply plantuml-maven-plugin/src/main/java/com/github/davidmoten/plantuml/plugins/SourceFileReaderExt.java]: I am also not happy about it, but this was the quick solution to mitigate the protected method problem. If you wish I can open a PR to PlantUML with change `getSuggestedFile` method visibility to public. If they accept it we can remove this class in a later release (I hope you will not mind to release with this workaround). What do you think?

**Review by davidmoten — APPROVED** [review]

**Comment by davidmoten** [comment]
Merged, thanks for the contribution. I'll release shortly.

**Comment by davidmoten** [comment]
[0.2.12](https://github.com/davidmoten/plantuml-maven-plugin/releases/tag/0.2.12) is on Maven Central now.

**Comment by balazs-toth-ing** [comment]
Thank you! Really appreciate your effort...

**Pull request merged**